### ISSUE TITLE
"PROTOCOL_ERROR" when using storage.googleapis.com, tokio-rustls-tls

### ISSUE THREAD

**Issue by bcspragu** [rustls/rustls#2108]
**Checklist**
* [x] I've searched the issue tracker for similar bugs.

**Describe the bug**

When using `rust-s3` with the `tokio-rustls-tls` backend, I reliably get this error when connecting to GCS:

```
Reqwest(reqwest::Error { kind: Request, url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("storage.googleapis.com")), port: None, path: "/<bucket>/rust-test", query: None, fragment: None }, source: hyper::Error(Http2, Error { kind: Reset(StreamId(1), PROTOCOL_ERROR, Remote) }) })
```

I initially filed this issue as https://github.com/durch/rust-s3/issues/377, but they (rightly) pointed me upstream. In my case, I was able to solve my issue by replacing `tokio-rustls-tls` with `tokio-native-tls`. This could of course be an issue with `tokio-rustls-tls`, but since using non-rustls fixes the issue, it seems this upstream is likely the culprit

**To Reproduce**
Copying the reproduction from https://github.com/durch/rust-s3/issues/377:

1. Have a Cargo.toml file with:

```
rust-s3 = { version = "0.33.0", default-features = false, features = ["tokio-rustls-tls"] }
```

2. Write a program like:
 
```rust
#[tokio::main]
async fn main() {
    let bucket_name = "<bucket>";
    let google_access_key_id = "<access key>";
    let google_access_key_secret = "<secret key>";

    let bucket = s3::bucket::Bucket::new(
        bucket_name,
        s3::region::Region::Custom {
            region: "auto".to_owned(),
            endpoint: "https://storage.googleapis.com".to_owned(),
        },
        s3::creds::Credentials::new(
            Some(google_access_key_id),
            Some(google_access_key_secret),
            None,
            None,
            None,
        )
        .unwrap(),
    )
    .unwrap()
    .with_path_style();

    bucket
        .put_object("rust-test", "some data".as_bytes())
        .await
        .unwrap();
}
```

**Applicable Version(s)**
This was reproduced on `rust-s3 v0.33.0`, which uses `rustls v0.21.12`, running on a recent Linux kernel, both inside and outside of Docker, built with Rust 1.75.

**Expected behavior**

For the example program above, I expected a file to be created on GCS at `/rust-test` containing `"some data"`, which happens correctly with `rust-s3`'s default features, but not with `tokio-rustls-tls`.

**Additional context**
As mentioned in the downstream bug https://github.com/durch/rust-s3/issues/377:

Googling around, it seems like Google servers do something with http2 that causes problems for `rustls`, but I didn't see anything conclusive.

**Comment by djc** [comment]
Would be nice if you can see about reproducing with rustls 0.23. FWIW I've definitely used the Google Cloud Storage REST API with rustls, although I just reqwest for it.

**Comment by bcspragu** [comment]
`rust-s3` is also using reqwest under the hood IIRC. I'll cut a version with 0.23 and see what happens.

**Comment by ctz** [comment]
My hunch would be this is something related to ALPN -- perhaps the protocols aren't being set by the client code, but then it is talking HTTP2 even though that wasn't negotiated?

**Comment by ctz** [comment]
With native-tls, no ALPN is offered which means HTTP2 is not used. With rustls, just "h2" is offered and the server agrees. 

Therefore we're actually testing HTTP/1.1 against HTTP2, rather than native-tls vs rustls.

The HTTP2 frames look like:

```
No.	Time	Source	Destination	Protocol	Length	Info
379	4.093822412	192.168.1.117	142.250.187.219	HTTP2	114	Magic
380	4.094172738	192.168.1.117	142.250.187.219	HTTP2	130	SETTINGS[0], WINDOW_UPDATE[0]
381	4.094469083	192.168.1.117	142.250.187.219	HTTP2	521	HEADERS[1]: PUT /%3Cbucket%3E/rust-test, DATA[1]
383	4.100608796	142.250.187.219	192.168.1.117	HTTP2	716	SETTINGS[0], WINDOW_UPDATE[0]
384	4.100608887	142.250.187.219	192.168.1.117	HTTP2	99	SETTINGS[0]
386	4.100821344	192.168.1.117	142.250.187.219	HTTP2	99	SETTINGS[0]
387	4.101612886	142.250.187.219	192.168.1.117	HTTP2	103	RST_STREAM[1]
388	4.101917767	192.168.1.117	142.250.187.219	HTTP2	107	GOAWAY[0]
```

Apparently, the HTTP2-level error is equivalent to a HTTP/1.1 400: https://github.com/hyperium/h2/issues/747 & https://github.com/summerwind/h2spec/issues/120

Exploring more:

- If I patch the rustls dependency to ignore ALPN, then the server returns exactly the same 400 response as native-tls.
- If I enable the `native-tls-alpn` feature on reqwest, I get the identical HTTP2 error as rustls.

So, in conclusion, there are several things going wrong here:

1. There is missing abstraction somewhere (reqwest?) between HTTP2 and HTTP/1.1 bad request errors, meaning libraries like rust-s3 need to be ready for both kinds of failure depending which HTTP version was negotiated.
2. Something is failing to use ALPN with native-tls, meaning HTTP2 is impossible with it (probably `rust-s3` crate failing to set `native-tls-alpn` feature on `reqwest`?)

Thanks for the good bug report, but since neither of those are issues with rustls, I will close this.
